besu/src/main/java/org/hyperledger/besu/controller/ConsensusScheduleBesuControllerBuilder.java

```diff
@@ -0,0 +1,334 @@
+import org.hyperledger.besu.ethereum.eth.sync.state.SyncState;
+import org.hyperledger.besu.ethereum.eth.transactions.TransactionPool;
+import org.hyperledger.besu.ethereum.eth.transactions.TransactionPoolConfiguration;
+import org.hyperledger.besu.ethereum.mainnet.MutableProtocolSchedule;
+import org.hyperledger.besu.ethereum.mainnet.ProtocolSchedule;
+import org.hyperledger.besu.ethereum.mainnet.ScheduledProtocolSpec;
+import org.hyperledger.besu.ethereum.p2p.config.SubProtocolConfiguration;
+import org.hyperledger.besu.ethereum.storage.StorageProvider;
+import org.hyperledger.besu.ethereum.worldstate.DataStorageConfiguration;
+import org.hyperledger.besu.ethereum.worldstate.PrunerConfiguration;
+import org.hyperledger.besu.ethereum.worldstate.WorldStateArchive;
+import org.hyperledger.besu.evm.internal.EvmConfiguration;
+import org.hyperledger.besu.metrics.ObservableMetricsSystem;
+import org.hyperledger.besu.plugin.services.permissioning.NodeMessagePermissioningProvider;
+
+import java.math.BigInteger;
+import java.nio.file.Path;
+import java.time.Clock;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Optional;
+
+import com.google.common.base.Preconditions;
+
+/*
+ This is a placeholder class for the QBFT migration logic. For now, all it does is to delegate any
+ BesuControllerBuilder to the first controller in the list.
+*/
+public class ConsensusScheduleBesuControllerBuilder extends BesuControllerBuilder {
+
+  private final Map<Long, BesuControllerBuilder> besuControllerBuilders = new HashMap<>();
+
+  public ConsensusScheduleBesuControllerBuilder(
+      final Map<Long, BesuControllerBuilder> besuControllerBuilderSchedule) {
+    Preconditions.checkNotNull(besuControllerBuilderSchedule);
+    Preconditions.checkArgument(!besuControllerBuilderSchedule.isEmpty());
+    this.besuControllerBuilders.putAll(besuControllerBuilderSchedule);
+  }
+
+  @Override
+  protected void prepForBuild() {
+    besuControllerBuilders.get(0L).prepForBuild();
+  }
+
+  @Override
+  protected MiningCoordinator createMiningCoordinator(
+      final ProtocolSchedule protocolSchedule,
+      final ProtocolContext protocolContext,
+      final TransactionPool transactionPool,
+      final MiningParameters miningParameters,
+      final SyncState syncState,
+      final EthProtocolManager ethProtocolManager) {
+    return besuControllerBuilders
+        .get(0L)
+        .createMiningCoordinator(
+            protocolSchedule,
+            protocolContext,
+            transactionPool,
+            miningParameters,
+            syncState,
+            ethProtocolManager);
+  }
+
+  @Override
+  protected ProtocolSchedule createProtocolSchedule() {
+    final MutableProtocolSchedule combinedProtocolSchedule =
+        new MutableProtocolSchedule(genesisConfig.getConfigOptions().getChainId());
+
+    for (Entry<Long, BesuControllerBuilder> builderSchedule : besuControllerBuilders.entrySet()) {
```

Comment:
Possibly can't guarantee iteration order with the HashMap?
Even if we changed the map impl to e.g. LinkedHashMap, is it still safer to sort by the keys first?

Reply:
Order we iterate in isn't important as long we can determine what spec is before and after. I've changed to the code in this area quite a bit though so my may not make sense with what was there before.